src/ExtMsg.actor.cpp

```diff
@@ -626,41 +626,36 @@ ACTOR Future<Reference<IReadWriteContext>> insertDocument(Reference<CollectionCo
                                                           bson::BSONObj d,
                                                           Optional<IdInfo> encodedIds) {
 	state Standalone<StringRef> encodedId;
-	state Standalone<StringRef> valueEncodedId;
 	state Optional<bson::BSONObj> idObj;
 	state Reference<QueryContext> dcx;
 
 	if (!encodedIds.present()) {
-		encodedId = DataValue(bson::OID::gen()).encode_key_part();
-		valueEncodedId = encodedId;
-		idObj = Optional<bson::BSONObj>();
+		// No _id field, generate a new one.
+		state bson::OID generatedId = bson::OID::gen();
+		state bson::BSONObj docWithGeneratedId =
+		    bson::BSONObjBuilder().appendElements(d).append("_id", generatedId).obj();
+		encodedId = DataValue(generatedId).encode_key_part();
 		dcx = cx->cx->getSubContext(encodedId);
+		// It's unlikely that the auto generated ID collides with an existing one, but just to be safe.
+		Optional<DataValue> existing = wait(dcx->get(DataValue("_id", DVTypeCode::STRING).encode_key_part()));
+		if (existing.present()) {
+			TraceEvent(SevError, "insertDocument: Auto generated object id collides with an existing ID")
+			    .detail("Auto generated ID", generatedId.toString());
+			throw duplicated_key_field();
+		}
+		// Insert doc first
+		insertElementRecursive("", docWithGeneratedId, dcx);
 	} else {
 		encodedId = encodedIds.get().keyEncoded;
-		valueEncodedId = encodedIds.get().valueEncoded;
-		idObj = encodedIds.get().objValue;
 		dcx = cx->cx->getSubContext(encodedId);
+
 		Optional<DataValue> existing = wait(dcx->get(DataValue("_id", DVTypeCode::STRING).encode_key_part()));
 		if (existing.present())
 			throw duplicated_key_field();
+		// Insert the doc
+		insertElementRecursive("", d, dcx);
 	}
 
-	// FIXME: abstraction violation out of laziness
-	Key prefix = StringRef(dcx->getPrefix().toString());
-	dcx->getTransaction()->tr->addWriteConflictRange(KeyRangeRef(prefix, strinc(prefix)));
```

Comment:
I don't know when these chunk was created but we don't need this explicit write conflict since all keys written will cause the respective write conflict added.

Reply:
If we are inserting a document like

```json
{
  "_id": "Xin",
  "section": "A"
}
```

Previous code would add write conflicts for the entire range starting with prefix `<coll_prefix>Xin`. Where without it, we will just add for individual keys `<coll_prefix>Xin`, `<coll_prefix>Xin:_id` and `<coll_prefix>Xin:section`. If there are any parallel transactions that are updating just any individual keys in between it could be dangerous. Most likely that shouldn't be an issue for inserts as it's not possible to have intermediate keys without `_id` field. And insert code checks for `_id` always.  Although, I am not entirely sure I am covering all the corner cases.

Adding write conflict range on the entire document space has two benefits

* It takes the conservative approach and covers all the corner cases.
* Although it is not obvious, adding the entire range is actually better for performance. It will create one conflict range for the insert. Whereas adding individual keys would add multiple conflict range, as each key converts into a range of single key, generating more load on resolver.

Reply:
I will add it back for now.